sfge/src/main/java/com/salesforce/rules/usewithsharingondatabaseoperation/UseWithSharingOnDatabaseOperationVisitor.java

```diff
@@ -0,0 +1,446 @@
+
+    /**
+     * For a more in-depth explanation, see {@link #afterVisit(DmlInsertStatementVertex,
+     * SymbolProvider)}
+     */
+    @Override
+    public void afterVisit(DmlUndeleteStatementVertex vertex, SymbolProvider symbols) {
+        createViolationIfDatabaseOpUnsafe(vertex, symbols);
+    }
+
+    /**
+     * For a more in-depth explanation, see {@link #afterVisit(DmlInsertStatementVertex,
+     * SymbolProvider)}
+     */
+    @Override
+    public void afterVisit(DmlUpsertStatementVertex vertex, SymbolProvider symbols) {
+        createViolationIfDatabaseOpUnsafe(vertex, symbols);
+    }
+
+    /**
+     * For a more in-depth explanation, see {@link #afterVisit(DmlInsertStatementVertex,
+     * SymbolProvider)}
+     */
+    @Override
+    public void afterVisit(DmlUpdateStatementVertex vertex, SymbolProvider symbols) {
+        createViolationIfDatabaseOpUnsafe(vertex, symbols);
+    }
+
+    @Override
+    public void afterVisit(SoqlExpressionVertex vertex, SymbolProvider symbols) {
+        createViolationIfDatabaseOpUnsafe(vertex, symbols);
+    }
+
+    /**
+     * @return the set of Violations compiled from all the post-visits
+     */
+    Set<Violation.PathBasedRuleViolation> getViolations() {
+        return violations;
+    }
+
+    private void createViolationIfDatabaseOpUnsafe(SFVertex vertex, SymbolProvider symbols) {
+        // afterVisit on all Database Operation vertices calls this method, but we only want to
+        // log a violation if this is the sink vertex and it's unsafe
+        if (vertex != null && vertex.equals(sinkVertex)) {
+            final SharingPolicyBoundary currentBoundary =
+                    sharingPolicyBoundaryDetector.getCurrentSharingBoundary();
+            final SharingPolicyBoundary effectiveBoundary =
+                    sharingPolicyBoundaryDetector.getEffectiveSharingBoundary();
+
+            SharingPolicy policy =
+                    effectiveBoundary.getApplicableAncestorPolicyVertex().getSharingPolicy();
+
+            switch (policy) {
+                case WITH_SHARING:
+                case INHERITED_SHARING:
+                    // we know that the operation runs as "with sharing" so this is not a violation,
+                    // however we need to check if the policy is implicitly inherited by the current
+                    // class and throw a warning if so.
+                    createWarningIfApplicable(currentBoundary, effectiveBoundary);
+                    // no need to check if a warning was created. The effective policy is safe, so
+                    // we just need to flag a warning if applicable.
+                    return;
+
+                case OMITTED_DECLARATION:
+                    // if there is a warning created, we know that this ends up running as "with
+                    // sharing", and so we should not create a violation. Otherwise, do.
+                    if (createWarningIfApplicable(currentBoundary, effectiveBoundary)) return;
```

Comment:
No brace-less `if`s. Also, might it be beneficial to split `createWarningIfApplicable` into `isWarningApplicable` and `createWarning`?